ocaml/xapi/xapi_clustering.ml

```diff
@@ -154,6 +154,24 @@ let assert_operation_host_target_is_localhost ~__context ~host =
   if host <> Helpers.get_localhost ~__context then
     raise Api_errors.(Server_error (internal_error, [ "A clustering operation was attempted from the wrong host" ]))
 
+(* a new member of a cluster should reside on the same network as the cluster *)
+let assert_same_subnet ~__context ~cluster ~pIF =
+  let parse_ip ip_str =
+    match Ipaddr.of_string ip_str with
+    | Ok ip_addr       -> ip_addr
+    (* this shouldn't happen because we shouldn't be storing invalid ips *)
+    | Error (`Msg msg) -> raise Api_errors.(Server_error (internal_error, [Printf.sprintf "could not parse ip address (%s), msg: %s" ip_str msg]))
+  in
+  let ip_of_pif pIF = ip_of_pif (pIF, Db.PIF.get_record ~__context ~self:pIF) |> Cluster_interface.str_of_address |> parse_ip in
+  let netmask_of_pif pIF = Db.PIF.get_netmask ~__context ~self:pIF |> parse_ip in
+
+  let network = get_network_internal ~__context ~self:cluster in
+  let pifs = Db.Network.get_PIFs ~__context ~self:network |> List.cons pIF in
+  let ips = List.map ip_of_pif pifs in
+  let netmask = List.map netmask_of_pif pifs |> Helpers.intersect_ips_exn in
+  if not (Helpers.are_on_same_subnet_exn netmask ips) then
```

Comment:
What if you have a PIF with netmask 255.255.255.255 (i.e. a /32 address) and IP 192.168.0.2,
and another PIF with netmask 255.255.255.0 (i.e. a /24 address) and IP 192.168.0.3.
When you intersect the netmasks I think you get 255.255.255.0 and conclude they are in the same subnet.
I think you should have a function for intersecting netmasks (using OR), or if you want to use the one for IPs then you should first invert the netmask, do the and, and then invert again.

Perhaps it'd be easier to avoid intersecting in the first place: you have one `pIF` to check, which has an IP address: do a `List.for_all` and check that this IP is a member of all the subnets that already exist (using Ipaddr.Prefix.mem).
That would also avoid having to handle IPv4 vs IPv6 separately.

Reply:
Yes, I don't think that intersecting netmasks is the right thing to do. If the constraint is that hosts need to be on the same subnet, then that means that the network prefixes (derived from IP + netmask) for all hosts are identical.

Reply:
For the case of adding a new host to an existing cluster, you could assume that the current cluster host are already on the same subnet. So just pick one and compare its subnet to the new host's?